### ISSUE TITLE
ProtoFlux nodes get unpacked by multiple users sometimes.

### ISSUE THREAD

**Issue by art0007i** [Yellow-Dog-Man/Resonite-Issues#375]
### Describe the bug?

Whenever I unpack a large creation in a world with multiple people, sometimes the nodes will have multiple visuals generated at the same time.

### To Reproduce

1. Go into a crowded world
2. Unpack a large ProtoFlux creation (In my case I unpacked the root of my Pose Mode installer: `resrec:///U-art0007i/R-426e9c63-230e-4dab-bb52-a44c68b0c82c`)
3. Look at the ProtoFlux nodes, some of them should have some Z-Fighting, and look much brighter than usual, at least for a few seconds.

### Expected behavior

ProtoFlux is only unpacked by a single person, and never generates extra node visuals.

### Screenshots

![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/19620451/562521eb-9107-4eca-a70b-328506224c97)

### Resonite Version Number

2023.10.20.831

### What Platforms does this occur on?

Windows

### Log Files

[ROTHORN - 2023.10.20.831 - 2023-10-21 13_19_34.log](https://github.com/Yellow-Dog-Man/Resonite-Issues/files/13060554/ROTHORN.-.2023.10.20.831.-.2023-10-21.13_19_34.log)

**Comment by Nytra** [comment]
I have seen this happen. Although from what I remember it didn't happen in a large ProtoFlux creation. It may have been a small RedPrint pack though.

**Comment by JackTheFoxOtter** [comment]
From my experience this only happens when you grab a node's slot reference and unpack from there, I've personally never had it happen when unpacking by grabbing the reference of the parent slot of the nodes and unpacking it.

**Comment by Armored-Dragon** [comment]
I am not certain this is a related bug, however one of my friends found a really nasty instance of node duplication.
On their avatar they had a maxwell that had about four "if" nodes in a single place. For context there were four of us in the session. These four nodes seemingly were enough to spike CPU usage for all members in the instance up to 90%!

Once the nodes were removed and replace, CPU usage stabilized to a much more reasonable level.
Here is the item if you want to take a look: resrec:///U-Armored-Dragon/R-c403c325-f678-41d7-b764-a3563d890c53

**Comment by art0007i** [comment]
> From my experience this only happens when you grab a node's slot reference and unpack from there, I've personally never had it happen when unpacking by grabbing the reference of the parent slot of the nodes and unpacking it.

Yeah we found a consistent way to replicate it with nytra that uses this feature as well. I think why it happened with a large creation and unpacking root before for me, is because I'm guessing my game lagged for long enough that it half unpacked and then everyone unpacked the 2nd half of the nodes.

**Comment by Nytra** [comment]
The original post has been edited with the new way to replicate the bug.

**Comment by 989onan** [comment]
I would also like to comment that this kind of "Node UI" duplication bug also appears with timers. I remember specifically with the user field being driven and in my case i was driving it with host. though I'm not sure if I should make a separate report since it happens even in singleplayer. it only requires the timer btw nothing else. though everything after it also duplicates

**Comment by BlueCyro** [comment]
I've looked at how protoflux unpacks and while I forget some of the specifics, the general jist is that *everyone* has the code that generates the visuals, and anybody who didn't trigger the generation of the visuals checks if a visual already exists. This works out fine most of the time since the user already has sent that the node now has a visual over the network.

However, there's some bits of the code that try to show the visuals if certain conditions are met, which I imagine is meant to catch scenarios where protoflux visuals should be generated, but otherwise aren't.

The side effect of this is that if the condition is met for the visual to spawn, other clients won't know that someone else is already trying to spawn it, which results in everyone spawning it and repeatedly initialize any wires connected to it. This is traditionally why the protoflux wires themselves were so heavy to unpack - which is alleviated by the fact that they're lighter on geometry now - but still happens in any case.

I'd meant to post my findings sooner, but better late than never

**Comment by art0007i** [comment]
If you had anything more specific, it would be nice to know what those conditions are. I guess a place to look is the replication case that we found with Nytra because it's very consistent.

And the lag that is caused by many users unpacking something at the same time is the main reason I made this issue.

**Comment by Nytra** [comment]
I think the cause of this is that sometimes only a single node is unpacked in a node group but no other nodes, then the ProtoFluxWireProxy OnChanges happens which causes the rest of the nodes to be unpacked by all users

**Comment by shiftyscales** [comment]
Would @BlueCyro have interest in investigating this issue further?

**Comment by BlueCyro** [comment]
> Would [@BlueCyro](https://github.com/BlueCyro) have interest in investigating this issue further?

Certainly.

**Comment by BlueCyro** [comment]
After looking, the cause has made it's terrible presence known to me.

When you unpack protoflux visuals, your client will set up each node, attach the wire proxies, etc. This part is fine.

The problem appears to majorly be with the wire proxies themselves. Each wire proxy will try to ensure the visuals of nodes that it's connected, as well as attach wires to any connection points within the OnChanges callback. OnChanges doesn't run at the exact same time as the initial visual setup does however, but rather one or two updates after the fact.

It's too late though, because your client has now sent out that there are visuals in the world without wires, and the wires that you've generated are still on their way to be synced. So now that everyone has registered that there are visuals without wires, the OnChanges code from the wire proxies will take effect on each person's client, and **everyone** in the session has now triggered visual/wire generation. All of the incoming messages generate lots of corrections as each client fights over setting up the wires/visuals as a result.

The reason the visuals still appear okay most of the time when all is said and done is because each wire proxy has a reference to a wire manager, but that reference is an offshoot SyncRef type called a CleanupRef that will delete what it points to when a new reference is written into it. This ends up exacerbating the above problem by doing N create/destroy operations on the wire manager, which means N create/destroy operations on all of the wire meshes in a protoflux pack for N players currently occupying the session.

This generates a **lot** of extra stuff to process and undoubtedly makes the GC scream from generating and destroying large amounts of mesh data.

But finally: Since each wire proxy will attempt to ensure the visuals of it's connected nodes, any nodes that don't have visuals registered will be guaranteed to have N copies of their visuals for N players in the session.

**Comment by Redd56** [comment]
> After looking, the cause has made it's terrible presence known to me.
https://www.youtube.com/watch?v=GMETU0BZc7k
